## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
# import Pandas Libruary
import pandas as pd
# Define a formatting function to format values as currency. Only use for printing 
def format_currency(value):
    return '${:,.2f}'.format(value)

In [2]:
client_data_df = pd.read_csv('Resources/client_dataset.csv')
# client_data_df.head()

In [3]:
# View the column names. (4 points)
client_data_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function. (4 points)
client_data_df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data
print(client_data_df.dtypes)
client_data_df.describe(include='all')

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object


first     last        job       phone                     email  \
count         54639    54639      54639       54639                     54639   
unique          360      477        498         999                       999   
top     Christopher  Johnson  Osteopath  5016915923  jessica.reyes@grimes.net   
freq           1090     1164        631         220                       220   
mean            NaN      NaN        NaN         NaN                       NaN   
std             NaN      NaN        NaN         NaN                       NaN   
min             NaN      NaN        NaN         NaN                       NaN   
25%             NaN      NaN        NaN         NaN                       NaN   
50%             NaN      NaN        NaN         NaN                       NaN   
75%             NaN      NaN        NaN         NaN                       NaN   
max             NaN      NaN        NaN         NaN                       NaN   

           client_id      order_id  order_date    order_week    order_year  \
count   54639.000000  5.463900e+04       54639  54639.000000  54639.000000   
unique           NaN           NaN         150           NaN           NaN   
top              NaN           NaN  2023-04-19           NaN           NaN   
freq             NaN           NaN         538           NaN           NaN   
mean    54837.869416  5.470190e+06         NaN     11.359139   2022.993064   
std     25487.438231  2.599807e+06         NaN      7.023499      0.082997   
min     10033.000000  1.000886e+06         NaN      1.000000   2022.000000   
25%     33593.000000  3.196372e+06         NaN      6.000000   2023.000000   
50%     53305.000000  5.496966e+06         NaN     11.000000   2023.000000   
75%     78498.000000  7.733869e+06         NaN     17.000000   2023.000000   
max     99984.000000  9.998480e+06         NaN     52.000000   2023.000000   

               item_id     category        subcategory    unit_price  \
count            54639        54639              54639  54639.000000   
unique             500            5                 18           NaN   
top     MAZ00314-80-8M  consumables  bathroom supplies           NaN   
freq               148        23538               6424           NaN   
mean               NaN          NaN                NaN    136.267207   
std                NaN          NaN                NaN    183.873135   
min                NaN          NaN                NaN      0.010000   
25%                NaN          NaN                NaN     20.800000   
50%                NaN          NaN                NaN     68.310000   
75%                NaN          NaN                NaN    173.160000   
max                NaN          NaN                NaN   1396.230000   

           unit_cost   unit_weight           qty   line_number  
count   54639.000000  54639.000000  5.463900e+04  54639.000000  
unique           NaN           NaN           NaN           NaN  
top              NaN           NaN           NaN           NaN  
freq             NaN           NaN           NaN           NaN  
mean       99.446073      5.004116  5.702646e+02      2.979667  
std       133.164267      5.326599  1.879552e+04      2.436320  
min         0.010000      0.000000  0.000000e+00      0.000000  
25%        14.840000      1.450000  3.200000e+01      1.000000  
50%        49.890000      3.240000  6.800000e+01      3.000000  
75%       125.570000      6.890000  1.700000e+02      5.000000  
max       846.270000     46.430000  3.958244e+06      9.000000

In [6]:
# Correctly identify the category with the most entries. (4 points)
# print (client_data_df['category'].value_counts().head(3).to_string(header=False,dtype=False),'\n')
client_data_df['category'].value_counts().head(1)

category
consumables    23538
Name: count, dtype: int64

In [7]:
# For the category with the most entries, correctly identify the subcategory with the most entries. (5 points)
print (client_data_df.groupby(['category','subcategory']).size().nlargest(1).to_string(header=False))


consumables  bathroom supplies    6424


In [8]:
# Correctly identify the 5 clients with the most entries in the data. (5 points)
 
print (client_data_df.groupby(['first','last']).size().nlargest(5).to_string(header=False,dtype=False))

Jessica    Reyes      220
Angela     Everett    211
Bryan      Myers      209
Alexandra  Young      207
Kendra     Garrett    207


In [9]:
# Store the client ids of those top 5 clients in a list. (4 points)
top_client_ids = client_data_df.groupby(['client_id']).size().nlargest(5).index.tolist()
top_client_ids

[33615, 66037, 46820, 24741, 38378]

In [10]:
# Display the total units (the qty column) that the client with the most entries ordered. (4 points)
client_data_df[client_data_df['client_id'] == top_client_ids[0]]['qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty. (6 points)
client_data_df['subtotal'] = client_data_df.apply((lambda x: x['unit_price']*x['qty']),axis=1)
print (client_data_df[['unit_price','qty','subtotal']].head(2).to_string(index=False),'\n')


 unit_price  qty  subtotal
    1096.80  105 115164.00
      24.95   21    523.95 



In [12]:
# Create a column for shipping price.
# Assume a shipping price of $8 per pound # for orders over 50 pounds
# and $10 per pound for items 50 pounds or under. (6 points)
client_data_df['shipping_weight'] = client_data_df.apply(lambda row: row['unit_weight'] * row['qty'],axis=1)
client_data_df['shipping_cost']   = client_data_df.apply(lambda row: 7.00 * row['shipping_weight'] 
                                    if row['unit_weight'] * row['qty'] > 50
                                    else 10.00 * row['unit_weight'] * row['qty'],axis=1)
print(client_data_df[['unit_price','unit_weight', 'qty', 'shipping_weight','shipping_cost']].head(3).to_string(index=False))

 unit_price  unit_weight  qty  shipping_weight  shipping_cost
    1096.80         7.50  105           787.50        5512.50
      24.95         1.49   21            31.29         312.90
      13.52         1.68   39            65.52         458.64


In [13]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

sales_tax = 1.0925
client_data_df['line_total'] = round(client_data_df.apply(lambda row: ((row['subtotal']) + row['shipping_cost'])*sales_tax,axis=1),2)
print(client_data_df[['qty', 'subtotal','shipping_cost','line_total']].head(3))

   qty   subtotal  shipping_cost  line_total
0  105  115164.00        5512.50   131839.08
1   21     523.95         312.90      914.26
2   39     527.28         458.64     1077.12


In [14]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%. (6 points)
client_data_df['line_cost'] = client_data_df.apply(lambda row: row['unit_cost'] * row['qty'] + row['shipping_cost'], axis=1)
client_data_df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number   subtotal  shipping_weight  shipping_cost  line_total  \
0            1  115164.00           787.50        5512.50   131839.08   
1            0     523.95            31.29         312.90      914.26   
2            6     527.28            65.52         458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [15]:
# Create a column for the cost of each line using unit cost, qty, and shipping price 
# (assume the shipping cost is exactly what is charged to the client). (6 points)
client_data_df['line_profit'] = client_data_df.apply(lambda row: row['line_total'] -row['line_cost'] , axis=1)
print(pd.concat([client_data_df['order_id'], client_data_df[['line_cost', 'line_total', 'line_profit']]
                 .map(format_currency)], axis=1).head().to_string(index=False))

 order_id  line_cost  line_total line_profit
  8953482 $85,597.05 $131,839.08  $46,242.03
  8069089    $629.79     $914.26     $284.47
  1902144    $765.18   $1,077.12     $311.94
  9031802  $1,077.35   $1,543.57     $466.22
  1322274  $8,663.60  $11,364.40   $2,700.80


## Part 3: Confirm your work

You have email receipts showing that the line_total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

- Order ID 2742071 had a total price of \$152,811.89
- Order ID 2173913 had a total price of \$162,388.71
- Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Confirm that Order ID 2742071 had a total price of $152,811.89. (5 points)
# Confirm that Order ID 2173913 had a total price of $162,388.71. (5 points)
# Confirm that Order ID 6128929 had a total price of $923,441.25. (5 points)
for order_id in [2742071, 2173913, 6128929]:
    # Calculate the sum
    customer_order_total = client_data_df.loc[client_data_df['order_id'] == order_id]['line_total'].sum()
    # format and print results
    print (f'Order {order_id}  Total ${customer_order_total:.2f}')

Order 2742071  Total $152811.89
Order 2173913  Total $162388.71
Order 6128929  Total $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# Calculate the total revenue from each of the top 5 clients in Part 1. (5 points)
top_client_ids = client_data_df.groupby(['client_id']).size().nlargest(5).index.to_list()
for client_id in top_client_ids:
    # Calculate the sum
    customer_order_total = client_data_df.loc[client_data_df['client_id'] == client_id]['line_total'].sum()
    # format and print results
    print (f'Client {client_id}\tTotal\t${customer_order_total:.2f}')

Client 33615	Total	$8377308.52
Client 66037	Total	$10259514.79
Client 46820	Total	$9743794.36
Client 24741	Total	$82268892.02
Client 38378	Total	$12906550.87


In [18]:
# Create a summary DataFrame showing the totals for the top 5 clients with the following information: 
# total units purchased, total shipping price, total revenue, and total profit. 
# Sort by total profit. (5 points)

summary_df = client_data_df[client_data_df['client_id'].isin(top_client_ids)].groupby('client_id').agg(
    total_qty=('qty', 'sum'),
    total_shipping=('shipping_cost', 'sum'),
    total_revenue=('line_total', 'sum'),
    total_cost=('line_cost', 'sum'),
    total_profit=('line_profit', 'sum')
    ).reset_index().sort_values(by='total_profit', ascending=False)
summary_df


client_id  total_qty  total_shipping  total_revenue   total_cost  \
0      24741     239862      5126448.37    82268892.02  45688899.71   
2      38378      73667      3429455.40    12906550.87   9634720.98   
4      66037      43018      1395151.85    10259514.79   7004482.98   
3      46820      75768      1601448.84     9743794.36   7007191.64   
1      33615      64313      1828984.89     8377308.52   6175313.91   

   total_profit  
0   36579992.31  
2    3271829.89  
4    3255031.81  
3    2736602.72  
1    2201994.61

In [19]:
# Format the data and rename the columns to names suitable for presentation. Currency should be in millions of dollars. (5 points)
# Rename the columns to names suitable for presentation
display_df = summary_df.rename(columns={
    'client_id': 'Client ID',
    'total_qty': 'Units',
    'total_shipping': 'Shipping',
    'total_revenue': 'Total Revenue',
    'total_cost': 'Total Cost',
    'total_profit': 'Total Profit'
})
# Convert currency values to millions with formatting
columns_to_convert = ['Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']
for column in columns_to_convert:
    display_df[column] = ('$' + (display_df[column] / 1_000_000).round(2).astype(str) + 'M')
print (display_df.to_string(index=False))

 Client ID  Units Shipping Total Revenue Total Cost Total Profit
     24741 239862   $5.13M       $82.27M    $45.69M      $36.58M
     38378  73667   $3.43M       $12.91M     $9.63M       $3.27M
     66037  43018    $1.4M       $10.26M      $7.0M       $3.26M
     46820  75768    $1.6M        $9.74M     $7.01M       $2.74M
     33615  64313   $1.83M        $8.38M     $6.18M        $2.2M


## Write a brief 2-3 sentence summary of your findings. (10 points)

# Summary
## The top client category is consumables
## The tops subcategory of consumables is bathroom supplies
## The top client by profit is Kendra Garrett